### PR TITLE
Fix bug where multiple keychain entries would result in user persistence failure

### LINKED ISSUE #5906: Keychain duplicate item error causing new Anonymous user on every login. No way to recover.
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 11.4.1
  * Firebase SDK version: 6.17.0
  * Firebase Component: Auth
  * Component version: _____
  * Installation method: Unity Package

#### Steps to reproduce:

This is inside a Unity app using the Unity FirebaseAuthentication package but it seems to be an underlying issue with the ios sdk.

1. Launch app
2. Authenticate as Anonymous User. Get a new user every time.

Keychain seems to get in a bad state and there is nothing in place to self-heal.

 [Firebase/Auth][I-AUT000001] Error loading saved user when starting up: Error Domain=FIRAuthErrorDomain Code=17995 "An error occurred when accessing the keychain. The @c NSLocalizedFailureReasonErrorKey field in the @c NSError.userInfo dictionary will contain more information about the error encountered" UserInfo={FIRAuthErrorUserInfoNameKey=ERROR_KEYCHAIN_ERROR, NSLocalizedFailureReason=SecItemCopyMatching (0), NSLocalizedDescription=An error occurred when accessing the keychain. The @c NSLocalizedFailureReasonErrorKey field in the @c NSError.userInfo dictionary will contain more information about the error encountered}

Every time the app is launched, it throws this error and generates a new anonymous user.

We had three ios Devices get into this state across ios12 and 13. This happened after installing over existing builds.

We could not break another device so we are not sure the exact repro steps.

Other relevant logs on these devices:

default	13:29:34.333261-0700	myapp	6.17.0 - [Firebase/Auth][I-AUT000001] Error loading saved user when starting up: Error Domain=FIRAuthErrorDomain Code=17995 "An error occurred when accessing the keychain. The @c NSLocalizedFailureReasonErrorKey field in the @c NSError.userInfo dictionary will contain more information about the error encountered" UserInfo={FIRAuthErrorUserInfoNameKey=ERROR_KEYCHAIN_ERROR, NSLocalizedFailureReason=SecItemCopyMatching (0), NSLocalizedDescription=An error occurred when accessing the keychain. The @c NSLocalizedFailureReasonErrorKey field in the @c NSError.userInfo dictionary will contain more information about the error encountered}

default	13:29:36.404742-0700	securityd	myapp[36042]/1#6 LF=0 add Error Domain=NSOSStatusErrorDomain Code=-25299 "duplicate item O,genp,E9E4417A,L,cku,428GFQXU64.*,0,acct,svce,v_Data,musr,20200624202936.397718Z,61FABC15" UserInfo={NSDescription=duplicate item O,genp,E9E4417A,L,cku,428GFQXU64.*,0,acct,svce,v_Data,musr,20200624202936.397718Z,61FABC15}
default	13:29:36.411227-0700	securityd	replaced <genp,rowid=31667,cdat=2020-06-19 22:55:49 +0000,mdat=2020-06-24 19:03:11 +0000,desc=null,icmt=null,crtr=null,type=null,scrp=null,labl=null,alis=null,invi=null,nega=null,cusi=null,prot=null,acct=firebase_auth_1___FIRAPP_DEFAULT_firebase_user,svce=firebase_auth_1:579799840025:ios:40a90b55b467ad03,gena=null,data=0ff8:070000000ADC140A...|128fabe608540c89,agrp=428GFQXU64.com.company.myapp,pdmn=cku,sync=0,tomb=0,sha1=A3369DCE76CFBCBAB30781565EC981BA1D711938,vwht=null,tkid=null,v_Data=<?>,v_pk=D0D9A9F44FCB4579611A6189DBDE46697F443504,accc=310D300B0C0470726F740C03636B75,u_Tomb=null,musr=,UUID=5E5160C5-7E4C-4CD1-8BA3-C7132D746FCD,sysb=null,pcss=null,pcsk=null,pcsi=null,persistref=> in <SecDbConnection rw open>
default	13:29:36.411293-0700	securityd	    with <genp,rowid=31667,cdat=2020-06-19 22:55:49 +0000,mdat=2020-06-24 20:29:36 +0000,desc=null,icmt=null,crtr=null,type=null,scrp=null,labl=null,alis=null,invi=null,nega=null,cusi=null,prot=null,acct=firebase_auth_1___FIRAPP_DEFAULT_firebase_user,svce=firebase_auth_1:579799840025:ios:40a90b55b467ad03,gena=null,data=0ff7:070000000ADC140A...|5807bf426d040ba,agrp=428GFQXU64.com.company.myapp,pdmn=cku,sync=0,tomb=0,sha1=90BC964473771E35C33ABAF93056110B63AAC822,vwht=null,tkid=null,v_Data=<?>,v_pk=D0D9A9F44FCB4579611A6189DBDE46697F443504,accc=310D300B0C0470726F740C03636B75,u_Tomb=null,musr=,UUID=5E5160C5-7E4C-4CD1-8BA3-C7132D746FCD,sysb=null,pcss=null,pcsk=null,pcsi=null,persistref=> in <SecDbConnection rw open>
default	13:29:36.434689-0700	myapp	Signed in fJQ3u47Rh7RoZv81HPs6bP2oXga2
System.EventHandler:Invoke(Object, EventArgs)
Firebase.Auth.FirebaseAuth:<StateChangedFunction>m__0(FirebaseAuth)
System.Action`1:Invoke(T)
System.Action:Invoke()
Firebase.ExceptionAggregator:Wrap(Action)
Firebase.Auth.FirebaseAuth:ForwardStateChange(IntPtr, Action`1)
Firebase.AppUtil:PollCallbacks()
Firebase.Platform.FirebaseHandler:Update()
 
(Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 35)

### PR DESCRIPTION
**Pull request by morganchen12**
Fixes #5906.

### CODE REVIEW

File: FirebaseAuth/Sources/Storage/FIRAuthKeychainServices.m [hunk 1]
```diff
@@ -129,11 +131,18 @@ - (nullable NSData *)itemWithQuery:(NSDictionary *)query error:(NSError **_Nulla
...
-    if (items.count != 1) {
+    if (items.count == 0) {
       if (error) {
+        // The keychain query returned no error, but there were no items found.
         *error = [FIRAuthErrorUtils keychainErrorWithFunction:@"SecItemCopyMatching" status:status];
       }
       return nil;
+    } else if (items.count > 1) {
+      // More than one keychain item was found, all but the first will be ignored.
+      FIRLogError(
```
Comment (paulb777) [inline FirebaseAuth/Sources/Storage/FIRAuthKeychainServices.m:142]: Should this be a warning?

File: FirebaseAuth/Sources/Storage/FIRAuthKeychainServices.m [hunk 2]
```diff
@@ -129,11 +131,18 @@ - (nullable NSData *)itemWithQuery:(NSDictionary *)query error:(NSError **_Nulla
...
+          kFIRLoggerAuth, @"I-AUT000005",
+          @"Keychain query returned multiple results, all but the first will be ignored: %@",
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/Storage/FIRAuthKeychainServices.m:144]: Should the match limit of 2 above be removed, in order to get all the ignored results?

**Comment by rosalyntan** [comment]
Is there a way to delete the duplicate users from the keychain? (I'm still new to iOS so don't know exactly how the keychain works)

**Comment by morganchen12** [comment]
One auth session will be discarded. Based on my understanding of the code, the two auth sessions are actually the same user, one from the legacy keychain service (with no service specified) and one from the up-to-date keychain service.

The right thing to do is probably to delete the legacy entry if we encounter both a legacy and non-legacy keychain entry.

**Comment by morganchen12** [comment]
I've modified the code to always return the non-legacy item if it exists. The way keychain deletions work, I'm not sure it's practical to delete the old keychain value without making significant modifications to the class.

Given the relatively limited impact of the bug I'm ok with not removing the legacy entry in this method.

File: FirebaseAuth/Sources/Storage/FIRAuthKeychainServices.m (hunk 1 above)
Comment (morganchen12) [reply FirebaseAuth/Sources/Storage/FIRAuthKeychainServices.m]: Done.

**Comment by paulb777** [comment]
Fix CI , then LGTM

**Pull request merged**